checksum.go

```diff
@@ -0,0 +1,245 @@
+//
+// GNU-style:
+//  <checksum>  file1
+//  <checksum> *file2
+//
+// see parseChecksumLine for more detail on checksum file parsing
+func extractChecksum(u *url.URL) (*fileChecksum, error) {
+	q := u.Query()
+	v := q.Get("checksum")
+
+	if v == "" {
+		return nil, nil
+	}
+
+	vs := strings.SplitN(v, ":", 2)
+	switch len(vs) {
+	case 2:
+		break // good
+	default:
+		// here, we try to guess the checksum from it's length
+		// if the type was not passed
+		return newChecksumFromValue(v, filepath.Base(u.EscapedPath()))
+	}
+
+	checksumType, checksumValue := vs[0], vs[1]
+
+	switch checksumType {
+	case "file":
+		return checksumFromFile(checksumValue, u)
+	default:
+		return newChecksumFromType(checksumType, checksumValue, filepath.Base(u.EscapedPath()))
+	}
+}
+
+func newChecksum(checksumValue, filename string) (*fileChecksum, error) {
+	c := &fileChecksum{
+		Filename: filename,
+	}
+	var err error
+	c.Value, err = hex.DecodeString(checksumValue)
+	if err != nil {
+		return nil, fmt.Errorf("invalid checksum: %s", err)
+	}
+	return c, nil
+}
+
+func newChecksumFromType(checksumType, checksumValue, filename string) (*fileChecksum, error) {
+	c, err := newChecksum(checksumValue, filename)
+	if err != nil {
+		return nil, err
+	}
+
+	c.Type = strings.ToLower(checksumType)
+	switch c.Type {
+	case "md5":
+		c.Hash = md5.New()
+	case "sha1":
+		c.Hash = sha1.New()
+	case "sha256":
+		c.Hash = sha256.New()
+	case "sha512":
+		c.Hash = sha512.New()
+	default:
+		return nil, fmt.Errorf(
+			"unsupported checksum type: %s", checksumType)
+	}
+
+	return c, nil
+}
+
+func newChecksumFromValue(checksumValue, filename string) (*fileChecksum, error) {
+	c, err := newChecksum(checksumValue, filename)
+	if err != nil {
+		return nil, err
+	}
+
+	switch len(c.Value) {
+	case md5.Size:
+		c.Hash = md5.New()
+		c.Type = "md5"
+	case sha1.Size:
+		c.Hash = sha1.New()
+		c.Type = "sha1"
+	case sha256.Size:
+		c.Hash = sha256.New()
+		c.Type = "sha256"
+	case sha512.Size:
+		c.Hash = sha512.New()
+		c.Type = "sha512"
+	default:
+		return nil, fmt.Errorf("Unknown type for checksum %s", checksumValue)
+	}
+
+	return c, nil
+}
+
+// checksumsFromFile will download checksumFile that is a checksum for file
+// behind src.
+//
+// checksumsFromFile will try to guess the hashing algorithm based on content
+// of checksum file
+//
+// checksumsFromFile will only return checksums for files that match file
+// behind src
+func checksumFromFile(checksumFile string, src *url.URL) (*fileChecksum, error) {
+	checksumFileURL, err := urlhelper.Parse(checksumFile)
+	if err != nil {
+		return nil, err
+	}
+
+	tempfile, err := tmpFile("", filepath.Base(checksumFileURL.Path))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		os.Remove(tempfile)
+	}()
+
+	if err = GetFile(tempfile, checksumFile); err != nil {
+		return nil, fmt.Errorf(
+			"Error downloading checksum file: %s", err)
+	}
+
+	filename := filepath.Base(src.Path)
+	absPath, _ := filepath.Abs(src.Path)
```

Comment:
Is there a reason why errors are being ignored here? They should probably be returned, while path parsing errors are probably going to be rare mid-request like this, having them is a good idea anyway.

Reply:
So, for these 2 cases, the errors are ignored because we could still have a match with `filepath.Base(src.Path)` and then not even need to check against absolute or relative paths.

Reply:
Ah ! I just read this: 
```
// Pwd is the working directory for detection. If this isn't set, some
// detection may fail. Client will not default pwd to the current
// working directory for security reasons.
```

So it might be worth it to actually not use `filepath.Abs`

Reply:
I don't think anything here really is a case though for not tracking these errors:

> `filepath.Abs` uses `os.Getwd()` which can error when the syscall fails

System call errors should _definitely_ be logged - we should not be continuing on a broken system

> `filepath.Rel` can fail in some special cases

Again, these special cases are things that we should be tracking an failing on anyway.

My point is that there aren't any cases here where it seems that the `filepath` calls fail due to non-fatal issues processing the path (due to some sort of malformed path). If there was, we would want to type-switch or otherwise check the error for these specific cases so that the rest of them were not blocked from being reported.

> So it might be worth it to actually not use filepath.Abs because it calls os.Getwd

Why is `Getwd` such a problem here? I've seldom seen this call fail, so if it does fail it's probably got a good reason and it's possibly something that is going to keep the rest of go-getter from functioning anyway.

Reply:
Gotcha, will do this, you are totally right !